src/unix/iostream_unix.ml

```diff
@@ -27,10 +27,142 @@ end
 module Out : sig
   open Out
 
-  val of_unix_fd : Unix.file_descr -> t_seekable
+  val of_unix_fd : ?close_noerr:bool -> Unix.file_descr -> t_seekable
   (** Output stream directly writing into the given Unix file descriptor. *)
 end = struct
   open Out
 
-  let of_unix_fd fd : t_seekable = of_out_channel (Unix.out_channel_of_descr fd)
+  let of_unix_fd ?(close_noerr = false) fd : t_seekable =
+    object
+      method output buf i len =
+        let i = ref i in
+        let len = ref len in
+        while !len > 0 do
+          let n = Unix.write fd buf !i !len in
+          i := !i + n;
+          len := !len - n
+        done
+
+      method seek i = ignore (Unix.lseek fd i Unix.SEEK_SET : int)
+      method pos () : int = Unix.lseek fd 0 Unix.SEEK_CUR
+
+      method close () =
+        if close_noerr then (
+          try Unix.close fd with _ -> ()
+        ) else
+          Unix.close fd
+    end
+end
+
+(** Sockets
+
+    @since NEXT_RELEASE *)
+module Sock : sig
+  val of_unix_sock :
+    ?close_noerr:bool ->
+    Unix.file_descr ->
+    Iostream.In.t_with_timeout * Iostream.Out.t
+  (** [of_unix_sock sock] puts [sock] in non-blocking mode and
+      returns a pair [ic, oc] of streams. *)
+end = struct
+  open Iostream
+
+  let now_ = Unix.gettimeofday
+  let short_timeout_ : float = 2.
+
+  let of_unix_sock ?(close_noerr = false) (fd : Unix.file_descr) :
+      In.t_with_timeout * Out.t =
+    let closed = Atomic.make false in
+
+    let close () =
+      if not (Atomic.exchange closed true) then
+        if close_noerr then (
+          try Unix.close fd with _ -> ()
+        ) else
+          Unix.close fd
+    in
+
+    Unix.set_nonblock fd;
+
+    let output_once buf i len =
+      let n = ref 0 in
+      while
+        (not (Atomic.get closed))
+        &&
+        try
+          n := Unix.write fd buf i len;
+          false
+        with
+        | Unix.Unix_error ((Unix.EAGAIN | Unix.EWOULDBLOCK), _, _) ->
+          (* sleep *)
+          true
+        | Unix.Unix_error ((Unix.ECONNRESET | Unix.ESHUTDOWN | Unix.EPIPE), _, _)
```

Comment:
the other exceptions I'd let bubble down, these are specifically handled here because they're the ones I can think of that related to either blocking, or the FD being actually closed.